From 0a3b37a9b703dcc5ddacb87d425196d9aa20b849 Mon Sep 17 00:00:00 2001
From: limeidan <limeidan@loongson.cn>
Date: Thu, 25 Aug 2022 11:14:15 +0800
Subject: [PATCH 35/48] [release-branch.go1.20] cmd/internal/sys, cmd/dist,
 misc/cgo/testcshared: enable c-shared feature and test on loong64

Linux kernel on loong64 has no Dup2 syscall support, so we use Dup3 to replace it like arm64 and riscv64.

Updates #53301
Fixes #58784

Change-Id: I4e0be140a71b86f4626ed39d76cf3ac78f842018
Reviewed-on: https://go-review.googlesource.com/c/go/+/425478
Auto-Submit: Ian Lance Taylor <iant@google.com>
Reviewed-by: Ian Lance Taylor <iant@google.com>
Run-TryBot: WANG Xuerui <git@xen0n.name>
Reviewed-by: David Chase <drchase@google.com>
Reviewed-by: abner chenc <chenguoqi@loongson.cn>
Run-TryBot: Ian Lance Taylor <iant@google.com>
TryBot-Result: Gopher Robot <gobot@golang.org>
Reviewed-by: xiaodong liu <teaofmoli@gmail.com>
Reviewed-by: WANG Xuerui <git@xen0n.name>
(cherry picked from commit b3a194aad63d476543be27a935849bd8e3f9534b)
---
 misc/cgo/testcshared/testdata/libgo2/dup2.go | 3 ++-
 misc/cgo/testcshared/testdata/libgo2/dup3.go | 3 ++-
 src/cmd/dist/test.go                         | 2 +-
 src/internal/platform/supported.go           | 2 +-
 4 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/misc/cgo/testcshared/testdata/libgo2/dup2.go b/misc/cgo/testcshared/testdata/libgo2/dup2.go
index d343aa54d9..9da2fd807c 100644
--- a/misc/cgo/testcshared/testdata/libgo2/dup2.go
+++ b/misc/cgo/testcshared/testdata/libgo2/dup2.go
@@ -2,7 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux,!arm64,!riscv64 netbsd openbsd
+//go:build darwin || dragonfly || freebsd || (linux && !arm64 && !loong64 && !riscv64) || netbsd || openbsd
+// +build darwin dragonfly freebsd linux,!arm64,!loong64,!riscv64 netbsd openbsd
 
 package main
 
diff --git a/misc/cgo/testcshared/testdata/libgo2/dup3.go b/misc/cgo/testcshared/testdata/libgo2/dup3.go
index 459f0dc196..999bfeaf26 100644
--- a/misc/cgo/testcshared/testdata/libgo2/dup3.go
+++ b/misc/cgo/testcshared/testdata/libgo2/dup3.go
@@ -2,7 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build linux,arm64 linux,riscv64
+//go:build (linux && arm64) || (linux && loong64) || (linux && riscv64)
+// +build linux,arm64 linux,loong64 linux,riscv64
 
 package main
 
diff --git a/src/cmd/dist/test.go b/src/cmd/dist/test.go
index e16a83db04..04c7fabcaa 100644
--- a/src/cmd/dist/test.go
+++ b/src/cmd/dist/test.go
@@ -1191,7 +1191,7 @@ func (t *tester) supportedBuildmode(mode string) bool {
 		return false
 	case "c-shared":
 		switch pair {
-		case "linux-386", "linux-amd64", "linux-arm", "linux-arm64", "linux-ppc64le", "linux-riscv64", "linux-s390x",
+		case "linux-386", "linux-amd64", "linux-arm", "linux-arm64", "linux-loong64", "linux-ppc64le", "linux-riscv64", "linux-s390x",
 			"darwin-amd64", "darwin-arm64",
 			"freebsd-amd64",
 			"android-arm", "android-arm64", "android-386",
diff --git a/src/internal/platform/supported.go b/src/internal/platform/supported.go
index 7350ee3120..d9e63e9cfb 100644
--- a/src/internal/platform/supported.go
+++ b/src/internal/platform/supported.go
@@ -143,7 +143,7 @@ func BuildModeSupported(compiler, buildmode, goos, goarch string) bool {
 
 	case "c-shared":
 		switch platform {
-		case "linux/amd64", "linux/arm", "linux/arm64", "linux/386", "linux/ppc64le", "linux/riscv64", "linux/s390x",
+		case "linux/amd64", "linux/arm", "linux/arm64", "linux/loong64", "linux/386", "linux/ppc64le", "linux/riscv64", "linux/s390x",
 			"android/amd64", "android/arm", "android/arm64", "android/386",
 			"freebsd/amd64",
 			"darwin/amd64", "darwin/arm64",
-- 
2.40.0

